dmriprep/interfaces/images.py

```diff
@@ -103,43 +91,279 @@ def _run_interface(self, runtime):
+    transforms = InputMultiObject(File(exists=True), mandatory=True)
 
-    img = nb.load(in_file)
-    if img.dataobj.ndim == 3:
-        return in_file
 
-    data = img.get_fdata(dtype='float32')
-    mask_img = nb.load(mask_file)
-    mask_data = mask_img.get_fdata(dtype='float32')
+class MatchTransformsOutputSpec(TraitedSpec):
+    transforms = OutputMultiObject(File(exists=True), mandatory=True)
+
+
+class MatchTransforms(SimpleInterface):
+    input_spec = MatchTransformsInputSpec
+    output_spec = MatchTransformsOutputSpec
+
+    def _run_interface(self, runtime):
+        self._results['transforms'] = match_transforms(self.inputs.dwi_files,
+                                                       self.inputs.transforms,
+                                                       self.inputs.b0_indices)
+        return runtime
+
 
-    median_signal = np.median(data[mask_data > 0, ...], axis=0)
-    rescaled_data = 1000 * data / median_signal
-    hdr = img.header.copy()
-    nb.Nifti1Image(rescaled_data, img.affine, hdr).to_filename(out_file)
-    return out_file
+class N3BiasFieldCorrection(ants.N4BiasFieldCorrection):
+    _cmd = "N3BiasFieldCorrection"
 
 
-def median(in_file, newpath=None):
-    """Average a 4D dataset across the last dimension using median."""
-    out_file = fname_presuffix(
-        in_file, suffix='_b0ref', newpath=newpath)
+class ImageMathInputSpec(BaseInterfaceInputSpec):
+    in_file = File(exists=True, mandatory=True, position=3, argstr='%s')
+    dimension = traits.Enum(3, 2, 4, usedefault=True, argstr="%d", position=0)
+    out_file = File(argstr="%s", genfile=True, position=1)
+    operation = traits.Str(argstr="%s", position=2)
+    secondary_arg = traits.Str("", argstr="%s")
+    secondary_file = File(argstr="%s")
 
-    img = nb.load(in_file)
-    if img.dataobj.ndim == 3:
-        return in_file
-    if img.shape[-1] == 1:
-        nb.squeeze_image(img).to_filename(out_file)
-        return out_file
 
-    median_data = np.median(img.get_fdata(dtype='float32'),
-                            axis=-1)
+class ImageMathOutputSpec(TraitedSpec):
+    out_file = File(exists=True)
 
-    hdr = img.header.copy()
-    hdr.set_xyzt_units('mm')
-    hdr.set_data_dtype(np.float32)
-    nb.Nifti1Image(median_data, img.affine, hdr).to_filename(out_file)
-    return out_file
+
+class ImageMath(CommandLine):
+    input_spec = ImageMathInputSpec
+    output_spec = ImageMathOutputSpec
+    _cmd = 'ImageMath'
+
+    def _gen_filename(self, name):
+        if name == 'out_file':
+            output = self.inputs.out_file
+            if not isdefined(output):
+                _, fname, ext = split_filename(self.inputs.in_file)
+                output = fname + "_" + self.inputs.operation + ext
+            return output
+        return None
+
+    def _list_outputs(self):
+        outputs = self.output_spec().get()
+        outputs['out_file'] = os.path.abspath(self._gen_filename('out_file'))
+        return outputs
```

Comment:
Look at that! https://github.com/poldracklab/niworkflows/blob/4e9181fd31f3a1b216b5edcf040bd5533c400e9d/niworkflows/interfaces/ants.py#L11-L77

Please use ImageMath from niworkflows (and PR against niworkflows if some functionality is missing)

Reply:
Awesome

Reply:
So the reason for redefining it outside of niworkflows is that we need it to be able to take an explicit secondary argument and file. In other words, as defined here, it is actually quite different. But we could monkey-patch it?

Reply:
Yes, we can import the niworkflows version and then add to it similar to the `DerivativesDataSink` and `BIDSDataGrabberOutputSpec` examples here: https://github.com/nipreps/dmriprep/blob/master/dmriprep/interfaces/__init__.py#L13-L20